### PR TITLE
feat(picker): Add volume picker to CellPicker

### PR DESCRIPTION
**Pull request by rodrigobasilio2022**
### Context

This PRs adds volume picker capabilities to Cellpicker Object. I recorded a video of the functionality working.

https://github.com/Kitware/vtk-js/assets/114958722/461c8878-f622-43c9-8ed5-1518cacedb43

### Changes

- [ ] Documentation and TypeScript definitions were updated to match those changes

### PR and Code Checklist

- [x] [semantic-release](https://github.com/semantic-release/semantic-release) commit messages
- [x] Run `npm run reformat` to have correctly formatted code

### Testing

- [x] This change adds or fixes unit tests 
- [x] Tested environment:
  - **vtk.js**: 
  - **OS**: 
  - **Browser**:

### CODE REVIEW

**Comment by floryst** [comment]
CI is failing due to old node versions. I will update it.

**Comment by rodrigobasilio2022** [comment]
Ok. Thanks. I tried to make fewer changes so you can review effectively. Do you want me to create a video?

**Comment by floryst** [comment]
If you have an example, that would be a great addition to demonstrate its capabilities.

One note I had skimming your changes is that the package.json changes are irrelevant to the functionality at hand. If the build works then I'm fine with the upgrades.

**Comment by rodrigobasilio2022** [comment]
THey are irrelevant. I wrongly executed npm install --force, that s it. I will make a video an put in the comments

**Comment by sedghi** [comment]
is this PR related to https://github.com/Kitware/vtk-js/pull/2885 ?

**Comment by floryst** [comment]
> is this PR related to #2885 ?

That appears to be a separate effort that hasn't been updated.

File: Sources/Rendering/Core/CellPicker/index.js [hunk 1]
```diff
@@ -244,6 +254,88 @@ function vtkCellPicker(publicAPI, model) {
...
 
+  publicAPI.intersectVolumeWithLine = (p1, p2, t1, t2, tol, actor, mapper) => {
+    let tMin = Number.MAX_VALUE;
+
+    const imageData = mapper.getInputData();
+    const origin = imageData.getOrigin();
+    const spacing = imageData.getSpacing();
+    const dims = imageData.getDimensions();
+    const scalars = imageData.getPointData().getScalars().getData();
+    const extent = imageData.getExtent();
```
Comment (finetjul) [inline Sources/Rendering/Core/CellPicker/index.js:265]: this should also consider the imageData `direction`.

File: Sources/Rendering/Core/CellPicker/index.js [hunk 2]
```diff
@@ -244,6 +254,88 @@ function vtkCellPicker(publicAPI, model) {
     return tMin;
   };
 
+  publicAPI.intersectVolumeWithLine = (p1, p2, t1, t2, tol, actor, mapper) => {
```
Comment (finetjul) [inline Sources/Rendering/Core/CellPicker/index.js:257]: maybe `actor, mapper` parameters could be grouped into a unique `volume` parameter instead.

File: Sources/Rendering/Core/Picker/index.js [hunk 3]
```diff
@@ -326,6 +330,32 @@ function vtkPicker(publicAPI, model) {
...
+        distance: vtkMath.distance2BetweenPoints(
+          p1World,
+          model.pickedPositions[i]
+        ),
+      });
+    }
+    tempArray.sort((a, b) => {
+      const keyA = a.distance;
+      const keyB = b.distance;
+      // Compare the 2 dates
```
Comment (finetjul) [inline Sources/Rendering/Core/Picker/index.js:348]: not dates.

Not sure this comment is useful anyway.

File: Sources/Rendering/Core/Picker/index.js [hunk 4]
```diff
@@ -326,6 +330,32 @@ function vtkPicker(publicAPI, model) {
       publicAPI.invokePickChange(model.pickedPositions);
       return 1;
     });
+    // sort array by distance
+    const tempArray = [];
+    for (let i = 0; i < model.pickedPositions.length; i++) {
+      tempArray.push({
+        actor: model.actors[i],
+        pickedPosition: model.pickedPositions[i],
+        distance: vtkMath.distance2BetweenPoints(
```
Comment (finetjul) [inline Sources/Rendering/Core/Picker/index.js:339]: rename distance into distance2

**Review by finetjul — COMMENTED** [review]
It's not clear why you had to modify package.json / package-lock.json

Please update CellPicker documentation accordingly.

File: Sources/Rendering/Core/CellPicker/index.js (hunk 1 above)
Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/CellPicker/index.js:265]: How do I add the information in the code?

File: Sources/Rendering/Core/CellPicker/index.js (hunk 2 above)
Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/CellPicker/index.js:257]: volumeActor is better? I can get the mapper form it

File: Sources/Rendering/Core/CellPicker/index.js (hunk 1 above)
Comment (finetjul) [reply Sources/Rendering/Core/CellPicker/index.js:265]: there must be some kind of a matrix inversion to apply to the line/ray I believe.

File: Sources/Rendering/Core/CellPicker/index.js (hunk 2 above)
Comment (finetjul) [reply Sources/Rendering/Core/CellPicker/index.js:257]: In VTK's vocabulary, volume == `vtkImageData`, actor == `vtkPolyData`, so I think `volume` is enough here (and you get the mapper from it like you suggested).

**Comment by rodrigobasilio2022** [comment]
Can you guys help me in git? I think I am doing a mess here. I want to delete the modifications I did in the package.json files

File: Sources/Rendering/Core/CellPicker/index.js (hunk 1 above)
Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/CellPicker/index.js:257]: I think p1 and p2 defines the ray and direction. That´s why always t, the parametric parameter is always positive between 0 and 1

File: Sources/Rendering/Core/CellPicker/index.js (hunk 2 above)
Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/CellPicker/index.js]: volume it is then...

File: Sources/Rendering/Core/CellPicker/index.js (hunk 1 above)
Comment (finetjul) [reply Sources/Rendering/Core/CellPicker/index.js:257]: Please create an example with headsq.vti, it has a non identity orientation.

**Review by finetjul — COMMENTED** [review]
Can you please also update index.d.ts files accordingly ?

**Comment by finetjul** [comment]
> Can you guys help me in git? I think I am doing a mess here. I want to delete the modifications I did in the package.json files

You will be able to squash all your branch commits eventually (I personnally use `git rebase -i` to do so)

File: Sources/Rendering/Core/CellPicker/index.js (hunk 1 above)
Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/CellPicker/index.js:265]: Add the example. Yes you are right!!!!

**Comment by rodrigobasilio2022** [comment]
Added the example. Something i noticed is why you guys d onot use the clipObject information
https://github.com/Kitware/vtk-js/blob/d61d2e44d98fa16803136754842a055ef2c41d21/Sources/Rendering/Core/CellPicker/index.js#L180

**Comment by finetjul** [comment]
> Added the example. Something i noticed is why you guys do not use the clipLine information in the following lines...
> 
> https://github.com/Kitware/vtk-js/blob/d61d2e44d98fa16803136754842a055ef2c41d21/Sources/Rendering/Core/CellPicker/index.js#L180

Not sure what do you mean, we do use `clipLine` afterwards

**Comment by rodrigobasilio2022** [comment]
The t1 and t2 updated values that restricts the ray

**Comment by finetjul** [comment]
> The t1 and t2 updated values that restricts the ray

I see, looks like a bug indeed. t1 and t2 should be extracted from the returned value...

File: Sources/Rendering/Core/CellPicker/index.js [hunk 5]
```diff
@@ -244,6 +254,126 @@ function vtkCellPicker(publicAPI, model) {
...
+        if (x[j] < extent[2 * j]) {
+          x[j] = extent[2 * j];
+          insideVolume = false;
+        } else if (x[j] > extent[2 * j + 1]) {
+          x[j] = extent[2 * j + 1];
+          insideVolume = false;
+        }
+
+        xi[j] = Math.floor(x[j]);
+        pcoords[j] = x[j] - xi[j];
```
Comment (finetjul) [inline Sources/Rendering/Core/CellPicker/index.js:354]: Is `pcoords` useful ?

File: Sources/Rendering/Core/CellPicker/index.js [hunk 6]
```diff
@@ -244,6 +254,126 @@ function vtkCellPicker(publicAPI, model) {
...
+        // Bounds check
+        if (x[j] < extent[2 * j]) {
+          x[j] = extent[2 * j];
+          insideVolume = false;
+        } else if (x[j] > extent[2 * j + 1]) {
+          x[j] = extent[2 * j + 1];
+          insideVolume = false;
+        }
+
+        xi[j] = Math.floor(x[j]);
```
Comment (finetjul) [inline Sources/Rendering/Core/CellPicker/index.js:353]: why not `round()` instead of `floor()`

File: Sources/Rendering/Core/Picker/index.js [hunk 7]
```diff
@@ -236,8 +236,11 @@ function vtkPicker(publicAPI, model) {
     props.forEach((prop) => {
       const mapper = prop.getMapper();
       pickable = prop.getNestedPickable() && prop.getNestedVisibility();
-      if (prop.getProperty().getOpacity() <= 0.0) {
-        pickable = false;
+
+      if (!mapper.isA('vtkVolumeMapper')) {
```
Comment (finetjul) [inline Sources/Rendering/Core/Picker/index.js:240]: if a volume has an opacity==0, it should probably not be pickable...

Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/Picker/index.js:240]: volumeMapper doesnt have this function

Comment (finetjul) [reply Sources/Rendering/Core/Picker/index.js:240]: I see, then you can simply do:
```
  if (prop.getProperty().getOpacity.?() <= 0.0) {
          pickable = false;
  }
```

Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/Picker/index.js:240]: if (prop.getProperty().getOpacity?.() <= 0.0) {
          pickable = false;
  }

Comment (rodrigobasilio2022) [reply Sources/Rendering/Core/Picker/index.js]: Done

File: Examples/Volume/VolumePicker/controlPanel.html [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+          <td>
+            Position
+            <input class='plane2Position' type="range" min="0" max="2.0" step="1" value="0" />
+          </td>
+          <td>
+            Rotation
+            <input class='plane2Rotation' type="range" min="0" max="2.0" step="1" value="0" />
+          </td>
+        </tr>
+    </tr>
```
Comment (finetjul) [inline Examples/Volume/VolumePicker/controlPanel.html:27]: Can you please expose the threshold opacity as a configurable parameter because in the example it is not obvious that it works as intended:
image
(I guess a property to control the opacity transfer function could be useful to demonstrate that it is unrelated to the threshold of the picker)

Comment (rodrigobasilio2022) [reply Examples/Volume/VolumePicker/controlPanel.html:27]: Is this you a looking for?
https://github.com/rodrigobasilio2022/vtk-js/blob/1f48bdc97ff6f73b4f137337b71604e7bcac45d9/Examples/Volume/VolumePicker/index.js#L188

Comment (finetjul) [reply Examples/Volume/VolumePicker/controlPanel.html:27]: Yes, having a slider to choose such threshold would be nice.

**Review by finetjul — APPROVED** [review]
LGTM

**Comment by floryst** [comment]
CI is red because this PR has some tests still failing. Please run `npm test` locally to identify the issue.

**Comment by rodrigobasilio2022** [comment]
I run the npm run test command and this was the final output


image

My machine is a Windows 11

**Comment by rodrigobasilio2022** [comment]
I executed the command again. In fact was executed yarn run test, and that was the output

image

**Comment by finetjul** [comment]
It looks like a test is hanging (that's why you do not see any failure)

**Comment by rodrigobasilio2022** [comment]
How do I solve this problem?

**Comment by rodrigobasilio2022** [comment]
Found the problem. It was a point picker function which not correspond with the changes in vtkPicker

**Comment by rodrigobasilio2022** [comment]
@finetjul I nned to do something else? I saw the PR was pulled out from merge queue, despite no conflict with the base branch

**Comment by floryst** [comment]
Can you rebase locally onto an updated master to see if there are any conflicts? Github recently had operational issues, so I don't know if there is an actual conflict or if the merge queues are currently broken.

**Comment by rodrigobasilio2022** [comment]
@floryst Made the change. It had one commit out from the base branch, despite no warnings in this PR. Please try again...

**Comment by rodrigobasilio2022** [comment]
@floryst @finetjul Now i think i got it right

**Review by finetjul — APPROVED** [review]
LGTM

**Comment by rodrigobasilio2022** [comment]
@finetjul I think the branch is ok right. It already passed teses tests?
